system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+ * Description:
+ *    Creates a setting for a float value
+ *
+ * Input Parameters:
+ *    setting        - pointer to the setting type
+ *    f              - the floating point value
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int set_float(FAR setting_t *setting, double f)
+{
+  DEBUGASSERT(setting);
+  if ((setting->type == SETTING_FLOAT) &&
+      (setting->type == SETTING_BYTE) &&
+      (setting->type == SETTING_INT))
+    {
+      return -EACCES;
+    }
+
+  setting->type = SETTING_FLOAT;
+  setting->val.f = f;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: get_ip
+ *
+ * Description:
+ *    Creates a setting for an IP address
+ *
+ * Input Parameters:
+ *    setting        - pointer to the setting type
+ *    ip             - pointer to return the IP address
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int get_ip(FAR setting_t *setting, FAR struct in_addr *ip)
+{
+  int ret = OK;
+  DEBUGASSERT(setting);
+  DEBUGASSERT((setting->type == SETTING_IP_ADDR) ||
+              (setting->type == SETTING_STRING));
+
+  if (setting->type == SETTING_IP_ADDR)
+    {
+      memcpy(ip, &setting->val.ip, sizeof(struct in_addr));
+    }
+  else if (setting->type == SETTING_STRING)
+    {
+      ret = inet_pton(AF_INET, setting->val.s, ip);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: set_ip
+ *
+ * Description:
+ *    Creates a setting for an IP address
+ *
+ * Input Parameters:
+ *    setting        - pointer to the setting type
+ *    ip             - IP address (in network byte order)
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int set_ip(FAR setting_t *setting, FAR struct in_addr *ip)
+{
+  DEBUGASSERT(setting);
+  if ((setting->type == SETTING_IP_ADDR) &&
+      (setting->type == SETTING_STRING))
+    {
+      return -EACCES;
+    }
+
+  setting->type = SETTING_IP_ADDR;
+  memcpy(&setting->val.ip, ip, sizeof(struct in_addr));
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: load
+ *
+ * Description:
+ *    Loads all values
+ *
+ * Input Parameters:
+ *    none
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+static int load(void)
+{
+  int i;
+  int ret = OK;
+
+  for (i = 0; i < CONFIG_SYSTEM_SETTINGS_MAX_STORAGES; i++)
+    {
+      if ((g_settings.store[i].file[0] != '\0') &&
+           g_settings.store[i].load_fn)
+        {
+          ret = g_settings.store[i].load_fn(g_settings.store[i].file);
+          if (ret < 0)
```

Comment:
OK, I get your point....but the load might fail for reasons the app might need to know about, such as the store is not valid, or a failed crc, meaning the app might want to do something about it.

Would you be happier/it be better if load was a void function, with the error and failed storage number returned via pointer parameters? Something else?

Reply:
I dunno... Typically the app doesn't bother with any IO. The app only cares about the RAM map (which never fails). IO is left to the system.

Even if you try to check the return value, a success does not necessarily mean success! (Due to driver or FS caching). So I simply ignored it, to keep things simple. I decided that it is best to work with the assumption that IO may have failed (and the app should be tolerable of this), rather than making false assumptions.

In any case, please don't break out of the loop. Give a chance to the rest of the storages.

Reply:
I agree we shouldn't bail out, and returning the load function return value for every store is a nuisance.

For now, I will remove the break and add a comment noting we are not sure what to do.

How about a new verify function? That could be called by the user app, doing a load that doesn't overwrite the RAM store but checks it's valid and crc integrity?

Reply:
Covered by a different comment now